Natalie Huang (#2)
bitprj/BitHeroesS20: blogp3.md

### Creating the React App

In the final part of these series, I'm going to be going over important steps in creating the frontend of our app in React. This tutorial won't cover the full code, so refer to [this](https://github.com/natalieh235/songrecproject/) github repo for all the deets.

### Setting up the Client in React

#### Step 1- Create the App

If you haven't already, use the command `npm install create-react-app -g` to install create-react-app globally on your machine. Then, get back into your Visual Studio Code root directory and use these commands to make a React app:

```
create-react-app client
cd client
```

You should now have two folders: server and client. 

Now you can run these commands to install your dependencies and start running your app.

```
npm install
npm start
```

Navigate to `http://localhost:3000` to see if your app is up and running. It should look something like this:

[![startingreactpage](https://github.com/natalieh235/songrecproject/raw/blog/images/startingreactpage.png)](https://github.com/natalieh235/songrecproject/blob/blog/images/startingreactpage.png)

#### Step 2- Create the Homepage

> So this React app will involve some very simple routing- basically this means that different urls will render different pages. For example, going to `localhost:3000/home` will take you to a homepage, while going to `localhost:3000/song` will perhaps show a list of songs. We're going to focus on creating a basic homepage, which will look something like this:

Reviewer: ```suggestion
So this React app will involve some very simple routing- this means that different urls will render different pages. For example, going to `localhost:3000/home` will take you to a homepage, while going to `localhost:3000/song` will perhaps show a list of songs. We're going to focus on creating a basic homepage, which will look something like this:
```

[![reacthomepage](https://github.com/natalieh235/songrecproject/raw/blog/images/reacthomepage.png)](https://github.com/natalieh235/songrecproject/blob/blog/images/reacthomepage.png)

> Get started by making a new file called `Home.js`. I stuck this inside a folder `components`, since we're going to end up making a lot more files later on.

Reviewer: ```suggestion
Get started by making a new file called `Home.js`. I stuck this inside a folder `components` since we're going to end up making a lot more files later on.
```

Add in this code:

```
import React from 'react';

function Home() {
  return (
    <div className="home-page">
        <h1>Upload your face. We pick the songs.</h1>
        <a href="https://localhost:8888">
          <button className="green-btn">Login with Spotify</button>
        </a>   
    </div>
  );
}

export default Home;
```

> This is a simple functional component, and simply contains a header and button. Notice that the button is redirecting us to our OAuth server page.

Reviewer: ```suggestion
This is a simple functional component, and it only contains a header and button. Notice that the button is redirecting us to our OAuth server page.
```

Return to your React app page and check to make sure the button takes you to the OAuth page, but don't log in quite yet.

#### Step 2- Creating the Redirect Page

> After the user logs in, we need to redirect them to a different page in our React app. To do this, I'm creating a new file called `UploadFace.js`. Note: this name ends up being kind of misleading, since this ends being the page where all the uploading and song-finding happens, but just bear with me.

Reviewer: ```suggestion
After the user logs in, we need to redirect them to a different page in our React app. To do this, I'm creating a new file called `UploadFace.js`. Note: this name ends up being kind of misleading since this ends being the page where all the uploading and song-finding happens, but just bear with me.
```

> Copy paste this code into UploadFace:

Reviewer: ```suggestion
Copy-paste this code into UploadFace:
```

> This creates a class component with state variables `token` and `userInfo`. Once the component mounts, we're using the lifecycle method `componentDidMount()` to immediately call the `getHashParams` function. The `getHashParams` function is actually used in the Spotify Oauth example code to parse the query string and extract the token. We're stealing this function to get the access token in our client-side.

Reviewer: ```suggestion
This creates a class component with state variables `token` and `userInfo`. Once the component mounts, we're using the lifecycle method `componentDidMount()` to immediately call the `getHashParams` function. The `getHashParams` function is used in the Spotify OAuth example code to parse the query string and extract the token. We're stealing this function to get the access token in our client-side.
```

> Once we get the token, we set our state variable `token` and make a API call using the token to get the user's display name and id. To test that this actually works, let's render a "Logged in as {display_name}" statement. This will change later.

Reviewer: ```suggestion
Once we get the token, we set our state variable `token` and make an API call using the token to get the user's display name and id. To test that this works, let's render a "Logged in as {display_name}" statement. This will change later.
```

```
import React from 'react';

class UploadFace extends React.Component {
  constructor(){
    super()
    
    this.state = {
      token: "",
      userInfo: []
    }
  }

  //calls getHashParams and uses token to get user info
  componentDidMount(){
    const params = this.getHashParams();
    this.setState({token: params.access_token})
  
    fetch('https://api.spotify.com/v1/me',{
      method: 'GET',
      headers: { 'Authorization' : 'Bearer ' + params.access_token}
    }).then(resp => resp.json()).then(data => this.setState({userInfo: [data.display_name, data.id]}))
    
  }

  //this parses the query string and extracts the token
  getHashParams() {
    console.log('in hash params')
    var hashParams = {};
    var e, r = /([^&;=]+)=?([^&;]*)/g,
        q = window.location.hash.substring(1);
    while ( e = r.exec(q)) {
       hashParams[e[1]] = decodeURIComponent(e[2]);
    }
    return hashParams;
  }

  render(){
    <h3>Logged in as {this.state.userInfo[0]}</h3>
  }
}

export default UploadFace;
```

Don't run this just yet- we have some routing to do.

#### Step 3- Adding Routing

> Go back into your App.js file. In your terminal run `npm install react-router-dom` . We'll be using react-router-dom to route our function, meaning that we'll be able to display different pages depending on the url.

Reviewer: ```suggestion
Go back into your App.js file. In your terminal run `npm install react-router-dom`. We'll be using react-router-dom to route our function, meaning that we'll be able to display different pages depending on the url.
```

> Copy paste this code:

Reviewer: ```suggestion
Copy-paste this code:
```

> Notice that I'm using `BrowserRouter`(saved as Router) and `Route` from the react-router-dom library. I'm surrounding all the components in my return statement with `<Router><Router/>`, which will allow me to specifiy actual Routes.

Reviewer: ```suggestion
Notice that I'm using `BrowserRouter`(saved as Router) and `Route` from the react-router-dom library. I'm surrounding all the components in my return statement with `<Router><Router/>`, which will allow me to specify Routes.
```

```
import React from 'react';
import './App.css';
import UploadFace from './components/UploadFace'
import Home from './components/Home'
import {BrowserRouter as Router, Route} from 'react-router-dom';

class App extends React.Component {

  render() {
    return (
      <Router>
        <div className="App">
          <Nav />
          <Route path="/face" component={UploadFace}/>
          <Route path="/" exact component={Home}/>
        </div>
      </Router>  
    );
  }
}

export default App;
```

These lines:

```
<Route path="/face" component={UploadFace}/>
<Route path="/" exact component={Home}/>
```

> are doing the actual routing. The path indicates the url that the user needs to go to for the page to render, and the component is the component that should be rendered at that specific url. Notice that there's a `<Nav />` component that isn't surrounded by route- this means that it will alway be rendered regardless of url. This is the navbar at the top of the page- you can customize this however you want, or eliminate it entirely.

Reviewer: ```suggestion
are doing the actual routing. The path indicates the url that the user needs to go to for the page to render, and the component is the component that should be rendered at that specific url. Notice that there's a `<Nav />` component that isn't surrounded by route- this means that it will always be rendered regardless of the given url. This is the navbar at the top of the page- you can customize this however you want, or eliminate it.
```

We're almost ready to test- the last step is actually to change the redirect url of your server. Find this line in your code(it should be around line 107):

```
res.redirect('/#' +
          querystring.stringify({
            access_token: access_token,
            refresh_token: refresh_token
          }));
```

This specifies the url that the access_token should be attached to. For us, it's going to be the React `UploadFace.js` page. Replace `'/#'` with `'http://localhost:3000/face/#'`. It's important to keep the hashtag because without it, our `getHashParams` function won't be able to parse the query string. Also notice that we are adding `/face` to our url- this changes the route so that we will see the UploadFace page instead of our homepage.

Make sure both your server and app are running(use `node app.js` for your server and `npm start` for your React app) and test it out!

[![loginexample](https://github.com/natalieh235/songrecproject/raw/blog/images/loginexample.gif)](https://github.com/natalieh235/songrecproject/blob/blog/images/loginexample.gif)

Yay it works! (hopefully)

#### Calling the Azure Function

> Remember the Azure Function we made all the way back in step one? Time to finally integrate it into our app. You may remember the test html page we made, and we're pretty much going to add that into our app. I created a component called `FaceForm.js`  with our html form inside:

Reviewer: ```suggestion
Remember the Azure Function we made in step one? Time to finally integrate it into our app. You may remember the test html page we made, and we're pretty much going to add that into our app. I created a component called `FaceForm.js`  with our html form inside:
```

```react
//FaceForm.js

import React, {useState} from 'react';
import FaceImage from './FaceImage'

function FaceForm(props){
    
    return(
        <div>
        
        <form encType="multipart/form-data" id="imageForm" >
          <div className="formPart">
          <label className="form-input" htmlFor="faceUploadInput">
            <input type="file" accept="image/*" name="image" id="faceUploadInput" onChange={(event) => props.loadFile(event)}/>
            <strong>Choose file</strong>
          </label>
           <div className="faceImage">
            <img alt='' id="facePic" src={props.img}></img>
        	</div>
          <button 
            type="submit"
            onClick={(event) => props.submitForm(event)} 
            className="green-btn"
            style={{display: props.showButton ? 'inline-block' : 'none' }}
            >Generate my playlist!</button>
          </div>
        </form>  
      </div>
    )
}

export default FaceForm
```

> This component is a child of `UploadFace.js` . Notice the event handler on the button has changed to call a function, `submitForm(event)`. The function gets the form data and calls our Azure Function using the function url. Let's take a look:

Reviewer: ```suggestion
This component is a child of `UploadFace.js`. Notice the event handler on the button has changed to call a function, `submitForm(event)`. The function gets the form data and calls our Azure Function using the function url. Let's take a look:
```

            
          </div>
          <SongPage emotions={this.state.emotions} token={this.state.token} id={this.state.userInfo[1]}/>
        </div>
      )
    }
    
    //if loading, show a loading page
    else if (this.state.loading){
      return <h2>Plz wait...</h2>
    }
    
    
    //if not submitted, show the image submission form
    return (
      <div>
        <UserProfile userInfo={this.state.userInfo}/>
        <FaceForm submitForm={this.submitForm} loadFile={this.loadFile} img={this.state.img} showButton={this.state.showFormButton}/>
      </div>
    )
  }
```

The loading page:

![loadingpage](/images/loadingpage.png)

 

**!!! IMPORTANT: YOU MUST SET YOUR REACT APP URL AS AN ALLOWED CORS ORIGIN IN YOUR AZURE FUNCTION**(refer to part 1 of this blog).

![CORS](images/CORS.png)

(Notice how http://localhost:3000 is an allowed origin)

#### Getting Song Recommendations from Spotify

Now that we've integrated our Azure Function into the React App, the rest of the process is pretty simple. I won't be going super in-depth into the code, just a few key points. The actual function that gets song recommendations is the `getRecommendations()` function. 

> A basic overview: the Spotify web API has a feature that allows you to call an API endpoint, enter in seed artists/seed tracks as params, and get a list of recommended tracks in return. In this function, I'm calling a different function called `getTop` , which uses the Spotify API to get a user's top tracks and artists.

Reviewer: ```suggestion
A basic overview: the Spotify web API has a feature that allows you to call an API endpoint, enter in seed artists/seed tracks as params, and get a list of recommended tracks in return. In this function, I'm calling a different function called `getTop`, which uses the Spotify API to get a user's top tracks and artists.
```

![finishedpage](/images/finishedpage.png)

> ### Adding a Webcam!

Reviewer: ```suggestion
### Bonus: Adding a Webcam!
```

> In an unplanned detour from my original project, I discovered this great npm library called [react-webcam](https://www.npmjs.com/package/react-webcam) that allows you easily add a webcam component in React. Naturally, I decided to add this in(though it's completely optional). This webcam will replace the image upload page we previously made. The full code can be found in the [webcam](https://github.com/natalieh235/songrecproject/tree/webcam) branch of the original repo.

Reviewer: ```suggestion
In an unplanned detour from my original project, I discovered this great npm library called [react-webcam](https://www.npmjs.com/package/react-webcam) that allows you easily add a webcam component to React. Naturally, I decided to add this in (though it's completely optional). This webcam will replace the image upload page we previously made. The full code can be found in the [webcam](https://github.com/natalieh235/songrecproject/tree/webcam) branch of the original repo. There are only a few steps to accomplish this goal.
```

> The code is pretty simple. First, I created a new React Component called `Cam.js`  for our webcam. [Here](https://github.com/natalieh235/songrecproject/blob/webcam/client/src/components/FormFolder/Webcam.js) is the complete `Cam.js` code. In this blog, I'll be breaking down in the code into individual pieces:

Reviewer: ```suggestion
First, I created a new React Component called `Cam.js`  for our webcam. [Here](https://github.com/natalieh235/songrecproject/blob/webcam/client/src/components/FormFolder/Webcam.js) is the complete `Cam.js` code. I'll be breaking down in the code into individual pieces:
```

> First, all of our import statments, including the one for `react-webcam`.

Reviewer: ```suggestion
Import all of the required packages, including the one for `react-webcam`.
```

```js
//Cam.js

import React from "react";
import Webcam from "react-webcam";
import notfound from './imagenotfound.png'
```

> Then, we declare our class component, called `Cam`, and the constructor:

Reviewer: ```suggestion
Then, we declare our class component, called `Cam`, and the constructor, noticing that the class will receive a `props` parameter:
```

> Notice that this class will receive a `props` parameter.

Reviewer: ```suggestion
```

```

The last part of this component is the render method:

The entire render code is wrapped in the `image-submission`  div, and then split into two major parts: `webcam-div` and `screenshots`.  

This is the `webcam-div` code, which renders the actual webcam with `<Webcam />`.  Notice that we have added a reference to the webcam with the variable `ref`.  We're going to need this later when we capture and screenshot. Also make sure you specify that `screenshotFormat='image/jpeg'`. **Without this, your image will be incorrectly encoded.**

```js
//Cam.js

render() {
      return (
        <div id='image-submission'>
          
          <div id="webcam-div">
            <h1>Take a picture:</h1>
            <Webcam
              audio={false}
              ref={node => this.webcam = node}
              screenshotFormat='image/jpeg'
              mirrored={true}
              style={{width: '100%'}}
            />
            <button 
              className="green-btn" 
              onClick={(event) => this.props.handleCapture(this.webcam.getScreenshot())}
              style={{width: "50%" }}
              >Capture
            </button>
          </div>
      );
    }
```

> We also have a `Capture`  button that is responsible for taking a screenshot whenever pressed. Notice that once the button is clicked, it's calling the `handleCapture`  that's been passed down from the `props` parameter. Inside the `handleCapture`  parameter, we are getting the base64 encoded string of the image by calling `this.webcam.getScreenshot()`. 

Reviewer: ```suggestion
We also have a `Capture`  button that is responsible for taking a screenshot whenever pressed. Notice that once the button is clicked, it's calling the `handleCapture` that's been passed down from the `props` parameter. Inside `handleCapture`, we are getting the base64 encoded string of the image by calling `this.webcam.getScreenshot()`. 
```

The `handleCapture` method is in the parent of the `Webcam`  component and looks like this:

> Essentially, we're updating the state variable `img` with the base64 string, and also indicating through updating `showFormButton` that a submit button should be rendered.

Reviewer: ```suggestion
Essentially, we're updating the state variable `img` with the base64 string, and also indicating that a submit button should be rendered through updating `showFormButton`.
```

```js
//UploadFace.js
handleCapture = (imgSrc) => {
    this.setState({ 
      img: imgSrc ,
      showFormButton: true
    });
  }
```

Now, onto the`screenshots` div, which makes up the second half of the `render`  method in `Cam.js`. This if statement: 

```js
{this.props.img ? <img src={this.props.img} /> : <img src={notfound} style={{width: '100%'}}/>}
```

> checks if there is an existing screenshot to display, and if not, displays a stock `image not found` photo.

Reviewer: ```suggestion
checks if there is an existing screenshot to display, and if not, it displays a stock `image not found` photo.
```

```js
//Cam.js
render(){
  ...
  
  <div id='screenshots'>
      <h1>Image</h1>
    
      {this.props.img ? <img src={this.props.img} /> : <img src={notfound} style={{width: '100%'}}/>}
    
      <button 
        className="green-btn" 
        onClick={(event) => this.props.submit(event)}
        style={{display: this.props.showButton ? 'inline-block' : 'none', width: "50%"}}
        >Submit
      </button>
    
    </div>
  </div>
}

```

> There's also a **Submit** button that calls the method`submit(event)`, which is also located in the parent component.

Reviewer: ```suggestion
There's also a **Submit** button that calls the method`submit(event)`, which is located in the parent component.
```

> The `submit` method sends the base64 encoded image string to our Azure function. Then, it checks if the response array has been filled(indicating a face has been recognized and analyzed). If the submission is invalid, the user has to retake the picture.

Reviewer: ```suggestion
The `submit` method sends the base64 encoded image string to our Azure function. Then, it checks if the response array has been filled (indicating a face has been recognized and analyzed). If the submission is invalid, the user has to retake the picture.
```

```js
async submit(event){  
  
  	//show the loading page
    this.setState({loading: true})
    event.stopPropagation();
    event.preventDefault();
  
    var payload = this.state.img;

  //calls the Azure Function
    const resp = await fetch("https://songrecapp.azurewebsites.net/api/SongRecTrigger", {
      method: 'POST',
      body: payload
    })

    var data = await resp.json();
    
  	//check if the data received is valid(a face has been recognized)
    if (data.result.length != 0){
      var emotions = data.result[0].faceAttributes.emotion;  
      console.log(emotions)
      this.setState(() => {
        return{
          emotions: emotions,
          submitted: true
        }
      }) 
      
    //if the resp.json() is empty, a face has not been recognized and they will stay on the page
    } else{
      this.setState(() => {
        return{
          loading: false
        }
      }) 
    }
  }
```

Recall that we previously sent our image in multipart-form data. Now, it's a base64 encoded string, which means that our Azure function will change slightly.

> The only different part is that we no longer have to parse the form data. Instead, we are decoding the base64 image string using the `Buffer.from()`  method, and then passing the new byte array into our `analyzeImage` function, which is exactly the same as before.

Reviewer: ```suggestion
The only different part is that we no longer have to parse the form data. Instead, we are decoding the base64 image string using the `Buffer.from()` method, and then passing the new byte array into the same `analyzeImage` function as before.
```
